design/20220118.server-side-apply.md

```diff
@@ -0,0 +1,167 @@
+One such example are Certificate SecretTemplate's Annotations and Labels. If a
+key was removed from either the Annotation or Label field on the SecretTemplate,
+the next reconciliation would not know which keys should be preserved or removed
+(as they may have been annotated or labelled from a third party). This problem
+occurs in several places where copying or transforming state from one resource
+to another happens.
+
+Managed fields define ownership of applied fields. cert-manager is therefore
+able to observe a discrepancy occurring from, for example, a previous
+SecretTemplate state and performing another apply (omitting those keys),
+resulting in those fields being removed on the Secret. Without managed fields,
+cert-manager would have to create a third state store in either annotations or
+status fields on that resource.
+
+## Field Manager and User Agent
+
+The Field Manager is a string denoting the manager who owns a field or set of
+fields on a resource. This string is set by the client at API call time and is
+derived by either the client setting the `fieldManager` URL query explicitly, or
+calculated from the client's user agent ([characters preceding the first "/",
+quote unprintable characters and then trim what's beyond the 128
+limit](https://github.com/kubernetes/kubernetes/blob/9a75e7b0fd1b567f774a3373be640e19b33e7ef1/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go#L252)).
+
+cert-manager will use a consistent naming scheme for both the user agent prefix
+and field manager across all components. Each component will have a human
+readable name that describes that component, and is used for both its user agent
+and field manager. Using the same name enables better auditing trails that helps
+debugging and improves telemetry.
+
+### Scheme
+
+### Field Manager
+
+The field manager is derived from the component in cert-manager, prefixed by the
+string `cert-manager`:
+
+```text
+cert-manager-<cert-manager component>
+```
+
+e.g.
+
+```text
+cert-manager-certificates-issuing
+```
+
+### User Agent
+
+The user agent is inspired by the [client-go default user
+agent](https://github.com/kubernetes/client-go/blob/664b1a6c8ce9c92ce65bef3f9833b402449c98d2/rest/config.go#L499),
+and is defined with the following:
+
+```text
+<component field manager>/<cert-manager version> (<Operating System>/<Architecture>) cert-manager/<git commit>
+```
+
+e.g:
+
+```text
+cert-manager-certificates-issuing/v1.7.0 (Linux/amd64) cert-manager/0b686b8f38c8c7442744c9224d18e780ee7f244a
+```
+
+## Implementation Consideration
+
+All CREATE and UPDATE operations in cert-manager controllers are to be replaced
+with APPLY. For this change, controllers will need to be modified so that for
+these calls, _only_ the fields in which they manage should be included in the
+PATCH API call. This means each controller will need to define exactly which
+fields they are concerned with.
```

Comment:
Can we discuss how this will be done in a non-intrusive way? (i.e. that permits/makes it easy to feature gate with minimal disruption to existing code paths?)